### DISCUSSION TITLE
Format levels for 3-level hierarchical HMM

### DISCUSSION THREAD

**Discussion by picardis** [bmcclintock/momentuHMM#156] [Q&A] [answered]
I am trying to fit a 3-level hierarchical HMM with data at 2 hour resolution for the coarse data stream (level 1), data at 15-minute resolution for the intermediate data stream (level 2), and data at 2-minute resolution for the fine data stream (level 3). I keep running into issues with the level assignment. From the vignette, “_every “1” observation must be followed by “2i”, every “2i” must be followed by one or more “2”, every “3i” must be preceded by “2” and followed by one or more “3”, and, after the first observation, every “1” must be preceded by an observation from level M._” However, I don’t understand how a “2i” can be followed by one or more “2” if the timestamps are nested across the three levels. I never have two consecutive intermediate-level points that are not separated by a bunch of fine-level points, because the intermediate data are at 15-minutes and the fine data are at 2-minutes. After each “2i” (a 15-minute resolution data point), I have a string of 2-minute data, i.e., a “3i” and a bunch of “3”. Then I switch back to “2i” at the next 15-minute mark, and then I immediately have 2-minute data again. To me, this seems similar to the fact that a “1” should not be followed by another “1” in a 2-level model, so I can’t figure out why a “2i” would be expected to be followed by a “2” in a 3-level model. What am I missing?

**Comment by bmcclintock** [comment, marked as answer]
The HHMM implementation in {momentuHMM} really isn't designed for this type of irregular/nested sampling design. `prepData` expects the order to be 1, 2i, 2, 3i, 3, ..., 3, 1, 2i, 2, 3i, 3, ..., 3, ....., 1, 2i, 2, 3i, 3, ..., 3. In other words, all coarser levels are expected to be observed (and/or be able to switch states) at the completion of each of the finer time series. In order to achieve what you want, you'd need to introduce additional rows in your data with the appropriate levels such that they conform to the expected pattern. 

If desired, you could use a dummy covariate to prevent state switching for the introduced observations at the coarser level(s). For example, `hierFormula` for level 1 could be `~dummy` (where `dummy=1` indicates when no switches can occur and `dummy=0` indicates when switches are allowed) and the `fixPar` argument could then be used to set the beta coefficients for the dummy covariate to a very large negative number. Building off of the cod example in the [vignette](https://cran.r-project.org/web/packages/momentuHMM/vignettes/momentuHMM.pdf), here state switches are forbidden for the second and third level 1 observations using a dummy covariate:

```
codData$dummy <- 0
codData$dummy[which(codData$level=="1")[2:3]] <- 1

hierFormula$AddChild("level1", formula=~dummy)

hierBeta$AddChild("level1",beta=matrix(c(-18.585, -2.86, -2.551, -1.641, -2.169, -2.415,rep(-1.e+10,6)),2,length(hierStates$children)*(length(hierStates$children)-1),byrow = TRUE))

fixPar <- list()
fixPar$beta <- data.tree::Node$new("cod HHMM fixPar")
fixPar$beta$AddChild("level1", fixPar=matrix(rep(c(NA,-1.e+10),each=length(hierStates$children)*(length(hierStates$children)-1)),2,byrow = TRUE))

checkPar0(codData,hierStates=hierStates,hierDist=hierDist,Par0=Par,hierFormula=hierFormula,DM=DM,hierBeta=hierBeta,hierDelta=hierDelta,fixPar=fixPar,estAngleMean = list(angle=TRUE))
### ----------------------------------------------------------
### Regression coeffs for the transition probabilities (beta):
### ----------------------------------------------------------
### -------------------------  level1  -----------------------
###                                1 -> 4    1 -> 7     4 -> 1     4 -> 7     7 -> 1     7 -> 4
### I((level == "1") * 1)     -1.8585e+01 -2.86e+00 -2.551e+00 -1.641e+00 -2.169e+00 -2.415e+00
### I((level == "1") * dummy) -1.0000e+10 -1.00e+10 -1.000e+10 -1.000e+10 -1.000e+10 -1.000e+10

iTrProbs <- getTrProbs(codData,hierStates=hierStates,hierBeta=hierBeta,hierFormula=hierFormula,hierDist=hierDist)
iTrProbs$level1$gamma[,,c(2,3)]
### , , 1
### 
###           resForage mobForage transit
### resForage         1         0       0
### mobForage         0         1       0
### transit           0         0       1
### 
### , , 2
### 
###           resForage mobForage transit
### resForage         1         0       0
### mobForage         0         1       0
### transit           0         0       1
```

**Reply by picardis** [reply]
Thanks, Brett! Is there an example of a 3-level HHMM anywhere that you are aware of? I am struggling to picture how the structure you described (1, 2i, 2, 3i, 3, ..., 3, 1, etc.) could arise from the sampling process. In a 2-level model, the fine-level data are nested within the coarse-level data so that each 1 is followed by a 2i and a string of 2. In a 3-level model, where the fine-level data are nested within intermediate-level data which are in turn nested within coarse-level data, why wouldn't each 2i be followed by a 3i just like every 1 is followed by a 2i? I think if I see an example this would become clear.

**Reply by bmcclintock** [reply]
Not that I'm aware of. With >2 levels there are a lot of potential states and parameters. You could perhaps try simulating a 3-level HHMM using `simHierData` to get a better feel for it (and fitting the simulated data using `fitHMM` to get a feel for the data requirements of such a complex model).
